ksql-common/src/main/java/io/confluent/ksql/function/GenericsUtil.java

```diff
@@ -0,0 +1,256 @@
+
+  private GenericsUtil() { }
+
+  /**
+   * @param typeName  the generic type name (e.g. {@code T})
+   * @return a {@link SchemaBuilder} for a generic type with that name
+   */
+  public static SchemaBuilder generic(final String typeName) {
+    return SchemaBuilder.bytes().optional().name(PREFIX + typeName + SUFFIX);
+  }
+
+  /**
+   * @param typeName  the generic type name (e.g. {@code T})
+   * @return a {@link SchemaBuilder} for a generic array with generic element type
+   */
+  public static SchemaBuilder array(final String typeName) {
+    return SchemaBuilder.array(generic(typeName).build()).optional();
+  }
+
+  /**
+   * @param keyType       the type for the map key
+   * @param valueTypeName the generic type name (e.g. {@code T}) for the map value
+   * @return a {@link SchemaBuilder} for a map with {@code keyType} keys and generic value type
+   */
+  public static SchemaBuilder map(final Schema keyType, final String valueTypeName) {
+    return SchemaBuilder.map(keyType, generic(valueTypeName).build()).optional();
+  }
+
+  /**
+   * @param type  the type
+   * @return whether or not {@code type} is a generic type
+   * @apiNote container generics (e.g. {@code ARRAY<T>}) will return {@code false},
+   *          use {@link #constituentGenerics(Schema)}
+   */
+  public static boolean isGeneric(final Schema type) {
+    return type.name() != null
+        && type.name().startsWith(PREFIX)
+        && type.name().endsWith(SUFFIX);
+  }
+
+  /**
+   * @param type  the type
+   * @return all generics contained within the type, for example: {@code Map<K, V>} would return
+   *         a set containing {@code {<K>, <V>}}
+   */
+  public static Set<Schema> constituentGenerics(final Schema type) {
+    switch (type.type()) {
+      case ARRAY:
+        return constituentGenerics(type.valueSchema());
+      case MAP:
+        return Sets.union(
+            constituentGenerics(type.keySchema()),
+            constituentGenerics(type.valueSchema()));
+      case STRUCT:
+        return type.fields().stream()
+            .map(Field::schema)
+            .map(GenericsUtil::constituentGenerics)
+            .flatMap(Collection::stream)
+            .collect(Collectors.toSet());
+      case BYTES:
+        if (isGeneric(type)) {
+          return ImmutableSet.of(type);
+        }
+        return ImmutableSet.of();
+      default:
+        return ImmutableSet.of();
+    }
+  }
+
+  /**
+   * Replaces all generics in a schema with concrete schemas defined in {@code mapping}
+   *
+   * @param schema  the schema which may contain generics
+   * @param mapping the mapping from generics to resolved types
+   * @return a schema with the same structure as {@code schema} but with no generics
+   *
+   * @throws KsqlException if there is a generic in {@code schema} that is not present
+   *                       in {@code mapping}
+   */
+  public static Schema resolve(final Schema schema, final Map<Schema, Schema> mapping) {
```

Comment:
nit: feel free to ignore this suggestion, but I would rename the functions in here this way:
`identifyGenerics` -> `resolveGenerics` - this is the function that's taking the generics and the provided schema and resolving what type the generic type actually is.
`resolve` -> `applyResolvedGenerics` - this takes the resolved generic types and applies that to the resolved schema.